File: daily/Backtracking/51-N-queens.py
class Solution:
    def solveNQueens(self, n: int) -> list[list[str]]:
        res = []
        cols, diag1, diag2 = set(), set(), set()
        board = [["."] * n for _ in range(n)]

        def backtrack(row):
            if row == n:
                current = []
                for r in board:
                    current.append("".join(r))
                res.append(current)
                return

            for col in range(n):
                if col in cols or (row - col) in diag1 or (row + col) in diag2:
                    continue
                board[row][col] = "Q"
                cols.add(col)
                diag1.add(row - col)
                diag2.add(row + col)
                backtrack(row + 1)
                board[row][col] = "."
                cols.remove(col)
                diag1.remove(row - col)
                diag2.remove(row + col)
        backtrack(0)
        return res

File: daily/Backtracking/51-N-queens.go
package daily

func solveNQueens(n int) [][]string {
	res := [][]string{}
	cols := map[int]bool{}
	diag1 := map[int]bool{}
	diag2 := map[int]bool{}

	// board[row][col], each cell is a byte '.' or 'Q'
	board := make([][]byte, n)
	for i := range board {
		board[i] = make([]byte, n)
		for j := range board[i] {
			board[i][j] = '.'
		}
	}

	var backtrack func(row int)
	backtrack = func(row int) {
		if row == n {
			current := make([]string, 0, n)
			for _, r := range board {
				current = append(current, string(r))
			}
			res = append(res, current)
			return
		}

		for col := 0; col < n; col++ {
			if cols[col] || diag1[row-col] || diag2[row+col] {
				continue
			}
			board[row][col] = 'Q'
			cols[col] = true
			diag1[row-col] = true
			diag2[row+col] = true

			backtrack(row + 1)

			board[row][col] = '.'
			delete(cols, col)
			delete(diag1, row-col)
			delete(diag2, row+col)
		}
	}

	backtrack(0)
	return res
}
